### DISCUSSION TITLE
Run vercel dev without linking

### DISCUSSION THREAD

**Discussion by garretcharp** [vercel/vercel#4925] [Help] [answered]
Hello, I am trying to figure out how to run `vercel dev` without linking a project?

I have seen a few other issues on this but all of them were related to CI/CD and they just needed to login which I do not want to have any login at all.

I just want to run `vercel dev` and it runs the server for the serverless functions.

Is this at all possible?

**Comment by styfle** [comment]
No, this is not possible. The project link is necessary to fetch the appropriate dev command. 

We will be begin autolinking during your first dev run in the next version of Vercel CLI.

**Comment by alexkvazos** [comment]
Why can't it just use vercel.json to figure this out? I don't understand the point of obligating any contributor to a codebase to having to create a vercel account to run a project, this is very aggresive.

**Reply by jadbox** [reply]
I also I'm considering on moving away from Vercel because this required login makes it difficult for both my team and automatic ux testing.

**Comment by leo** [comment, marked as answer]
Thank you for your feedback, everyone!

---

The reason why `vercel dev` currently requires login is because it retrieves a number of things from the Project Settings, such as Environment Variables defined for the "Development" Environment and the Development Command.

Some of those things could be hardcoded directly into the Git repository, some others not, as they might contain sensitive information, such as tokens.

Regardless, however, we recognize the need for wanting to be able develop with `vercel dev` as easily as possible and the problems that this requirement is creating for open source contributors, so we're looking to provide a solution in the coming months.

---

In the meantime, this might be helpful:

If you're using a framework, you likely don't even need to run `vercel dev`. Instead, you can just run the framework's own Development Command. The only case in which you'd need `vercel dev` is if you'd like to define custom routing rules or API routes and the framework does not support those features natively.

In the case of frameworks like Next.js or RedwoodJS, for example, there wouldn't even be a reason to use `vercel dev` because they support those features natively (Next.js supports both and RedwoodJS supports API routes).

**Reply by alexkvazos** [reply]
Thanks leo! Yes, when using Next.js its not a problem, but some times you just want to test out quick endpoint or something and not have to make a full-featured next app. Thank you for having this in your roadmap!

**Reply by jadbox** [reply]
I didn't know that you could also specify rewrites in next.config.js too! That's perfect- thank you!

**Reply by sauravhiremath** [reply]
@leo Just wanted to add another use case, where using `vercel dev` would be preferred over `next dev`

https://github.com/vercel/vercel/issues/4739#issuecomment-783787967

**Comment by homerjam** [comment]
This is rather annoying, I find my self having to avoid typescript and create a separate development setup. I don't even find the shared environment vars that useful personally, they could be optional surely?
